src/pocketmine/network/upnp/UPnP.php

```diff
@@ -19,63 +19,183 @@
+// LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
+// A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+// OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+// SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+// LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+// DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+// THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+// (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+// OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
 declare(strict_types=1);
 
 /**
- * UPnP port forwarding support. Only for Windows
+ * UPnP port forwarding support.
  */
 namespace pocketmine\network\upnp;
 
 use pocketmine\utils\Internet;
-use pocketmine\utils\Utils;
-use function class_exists;
-use function is_object;
+use SimpleXMLElement;
+use function socket_create;
+use function socket_set_option;
+use function socket_recvfrom;
+use function socket_sendto;
+use function trim;
+use function strlen;
+use function preg_match;
+use function parse_url;
+use const AF_INET;
+use const SOCK_DGRAM;
+use const SOL_UDP;
+use const IPPROTO_IP;
+use const IP_MULTICAST_IF;
+use const IP_MULTICAST_LOOP;
+use const IP_MULTICAST_TTL;
+use const SOL_SOCKET;
+use const SO_RCVTIMEO;
 
 abstract class UPnP{
+	/** @var string|null */
+	private static $serviceURL;
+	
+	public static function getControlUrl() : string{
+		try{
+			$socket = socket_create(AF_INET, SOCK_DGRAM, SOL_UDP);
+			socket_set_option($socket, SOL_SOCKET, SO_RCVTIMEO, ["sec" => 3, "usec" => 0]);
+			$contents =
+				"M-SEARCH * HTTP/1.1\r\n" .
+				"MX: 2\r\n" .
+				"HOST: 239.255.255.250:1900\r\n" .
+				"MAN: \"ssdp:discover\"\r\n" .
+				"ST: upnp:rootdevice\r\n\r\n";
+			socket_sendto($socket, $contents, strlen($contents), 0, "239.255.255.250", 1900);
+			socket_recvfrom($socket, $buffer, 1024, 0, $responseHost, $responsePort);
+			socket_close($socket);
+		}catch(\Exception $e){
+			throw new \RuntimeException("Socket error: " . $e->getMessage());
+		}
+		if($buffer === false || !preg_match('/location\s*:\s*(.+)\n/i', $buffer, $matches)){
+			throw new \RuntimeException("Unable to find the router. Ensure that network discovery is enabled in Control Panel.");
+		}
+		$location = trim($matches[1]);
+		$url = parse_url($location);
+		if($url === false){
+			throw new \RuntimeException("Failed to parse the router's url: {$location}");
+		}
+		$response = Internet::getURL($location, 3, [], $err, $headers, $httpCode);
+		if($response === false){
+			throw new \RuntimeException("Unable to access XML: {$err}");
+		}
+		if($httpCode !== 200){
+			throw new \RuntimeException("Unable to access XML: {$response}");
+		}
+
+		$defaultInternalError = libxml_use_internal_errors(true);
+		try{
+			$root = new \SimpleXMLElement($response);
+		}catch(\Exception $e){
```

Comment:
please check a more specific exception if possible

Reply:
`get_class($e)` returned 'Exception', so I think it raises Exception itself.

Reply:
That's fair enough. I just get uncomfortable seeing generalized catches like this.